### PR TITLE
Nullable type

### PR DESCRIPTION
**Pull request by gilbertw**

### CODE REVIEW

File: sdk/core/azure-core/inc/nullable.hpp [hunk 1]
```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <new> // for placement new
+#include <stdlib.h> // for abort
```
Comment (JinmingHu-MSFT) [inline sdk/core/azure-core/inc/nullable.hpp:7]:
```suggestion
#include <cstdlib> // for abort
```

**Comment by bhardwajs** [comment]
Wouldn't be better to have the implementation similar to `std::optional` and if C++ 17 is available, default to `std::optional`?
https://github.com/microsoft/STL/blob/master/stl/inc/optional

**Comment by JinmingHu-MSFT** [comment]
> Wouldn't be better to have the implementation similar to `std::optional` and if C++ 17 is available, default to `std::optional`?
> https://github.com/microsoft/STL/blob/master/stl/inc/optional

We cannot use C++17. The latest C++ standard we can use is C++14

**Comment by bhardwajs** [comment]
> 
> 
> > Wouldn't be better to have the implementation similar to `std::optional` and if C++ 17 is available, default to `std::optional`?
> > https://github.com/microsoft/STL/blob/master/stl/inc/optional
> 
> We cannot use C++17. The latest C++ standard we can use is C++14

[Bond](https://github.com/microsoft/bond) needs to support C++ 11 and up. There, I created a conditional wrapper to support C++ 17 conditional code: 
https://github.com/bhardwajs/bond/blob/e83de6ffe5fcb38fb3e9f637351f3f9b90479dde/cpp/inc/bond/core/config.h#L35

Similarly, this can be
```
#if CXX_17
using Nullable<t> = std::optional<t>;
#else
...
#endif
```

**Comment by JinmingHu-MSFT** [comment]
> > > Wouldn't be better to have the implementation similar to `std::optional` and if C++ 17 is available, default to `std::optional`?
> > > https://github.com/microsoft/STL/blob/master/stl/inc/optional
> > 
> > 
> > We cannot use C++17. The latest C++ standard we can use is C++14
> 
> [Bond](https://github.com/microsoft/bond) needs to support C++ 11 and up. There, I created a conditional wrapper to support C++ 17 conditional code:
> https://github.com/bhardwajs/bond/blob/e83de6ffe5fcb38fb3e9f637351f3f9b90479dde/cpp/inc/bond/core/config.h#L35
> 
> Similarly, this can be
> 
> ```
> #if CXX_17
> using Nullable<T> = std::optional<T>;
> #else
> ...
> #endif
> ```

Then we have to maintain two sets of code, which means extra work. I don't know if it's a good idea.

I'm fine with either way.

**Comment by katmsft** [comment]
> Wouldn't be better to have the implementation similar to `std::optional` and if C++ 17 is available, default to `std::optional`?
> https://github.com/microsoft/STL/blob/master/stl/inc/optional

I think @BillyONeal is the best person to answer this question.

File: CMakeSettings.json [hunk 2]
```diff
@@ -1,28 +1,28 @@
...
+            "buildRoot": "${projectDir}\\out\\build\\${name}",
+            "installRoot": "${projectDir}\\out\\install\\${name}",
+            "cmakeCommandArgs": "",
+            "buildCommandArgs": "",
+            "ctestCommandArgs": "",
+            "inheritEnvironments": [ "msvc_x86" ],
+            "variables": []
+        }
+    ]
 }
```
Comment (katmsft) [inline CMakeSettings.json:28]: nit: Add new line.

File: sdk/core/azure-core/inc/nullable.hpp (hunk 1 above)
Comment (gilbertw) [reply sdk/core/azure-core/inc/nullable.hpp:7]: > > > Wouldn't be better to have the implementation similar to `std::optional` and if C++ 17 is available, default to `std::optional`?
> > > https://github.com/microsoft/STL/blob/master/stl/inc/optional
> > 
> > 
> > We cannot use C++17. The latest C++ standard we can use is C++14
> 
> [Bond](https://github.com/microsoft/bond) needs to support C++ 11 and up. There, I created a conditional wrapper to support C++ 17 conditional code:
> https://github.com/bhardwajs/bond/blob/e83de6ffe5fcb38fb3e9f637351f3f9b90479dde/cpp/inc/bond/core/config.h#L35
> 
> Similarly, this can be
> 
> ```
> #if CXX_17
> using Nullable<T> = std::optional<T>;
> #else
> ...
> #endif
> ```

That is correct that we can't have C++17 dependencies.  Nullable<T> is slimmer version of optional<T> and not 100% compatible with std::optional<T>.  The above MACRO would imply either works in all situations and that is not true.

File: sdk/core/azure-core/inc/nullable.hpp [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+  public:
+    constexpr Nullable() : m_disengaged{}, m_hasValue(false) {}
+    constexpr Nullable(const T& initialValue) : m_value(initialValue), m_hasValue(true) {}
+
+    Nullable(const Nullable& other) noexcept(std::is_nothrow_copy_constructible<T>::value)
+        : m_hasValue(other.m_hasValue)
+    {
+      if (m_hasValue)
+      {
+        // standard implementation has to use std::addressof instead of & :)
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:40]: I left these comments for Jeffrey but you don't need them here.

File: sdk/core/azure-core/inc/nullable.hpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+      // this move and swap may be inefficient for some Ts but
+      // it's a lot less code than the standard implementation :)
+      Nullable{std::move(other)}.Swap(*this);
+      return *this;
+    }
+
+    template <
+        class U = T,
+        typename std::enable_if<
+            !std::is_same<Nullable, std::remove_cv_t<std::remove_reference_t<U>>>::value // Avoid repeated assignment
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:117]: This constraint is coped from a forwarding assignment operator but you're not forwarding here. 

All the _ts should not compile in C++14 mode.

File: sdk/core/azure-core/inc/nullable.hpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      }
+      else if (other.m_hasValue)
+      {
+        ::new (static_cast<void*>(&m_value)) T(std::move(other.m_value)); // throws
+        m_hasValue = true;
+        other.Destroy();
+      }
+    }
+
+    // Intentionally lowercase to override the swap
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:91]:
```suggestion
    // Intentionally lowercase to follow the Swappable requirements
```

https://en.cppreference.com/w/cpp/named_req/Swappable

File: sdk/core/azure-core/inc/nullable.hpp (hunk 4 above)
Comment (BillyONeal) [reply sdk/core/azure-core/inc/nullable.hpp:117]: (I'm extremely nervous about copying one overload of the std::optional overload set without copying every overload; that is a finely tuned 'mess')

File: sdk/core/azure-core/inc/nullable.hpp [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+      return *this;
+    }
+
+    bool HasValue() const noexcept { return m_hasValue; }
+
+    const T& GetValue() const& noexcept
+    {
+      if (!m_hasValue)
+      {
+        // throwing here prohibited by our guidelines
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:138]: This was a message for you / Jeffrey, I don't know if it adds value committed.

File: sdk/core/azure-core/inc/nullable.hpp [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+        // https://azure.github.io/azure-sdk/cpp_design.html#pre-conditions
+        std::abort();
+      }
+
+      return std::move(m_value);
+    }
+
+    explicit operator bool() const { return HasValue(); }
+
+    // GetValueOrDefault() does NOT make sense here given that T
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:172]: Note that you still might want to add a value_or function.

File: sdk/core/azure-core/inc/nullable.hpp [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+            !std::is_same<Nullable, std::remove_cv_t<std::remove_reference_t<U>>>::value // Avoid repeated assignment
+            && !(std::is_scalar<U>::value && std::is_same<T, std::decay_t<U>>::value) //Avoid repeated assignment of equivallent scaler types
+            && std::is_constructible<T, U>::value   //Ensure the type is constructible
+            && std::is_assignable<T, U>::value      //Ensure the type is assignable
+            && !std::is_null_pointer<U>::value,     //Dissallow nullptr assignment
+            int>::type
+        = 0>
+    Nullable& operator=(const U& other)
+    {
+      // this copy and swap may be inefficient for some Ts but
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:126]: Here you know there's no self-alias problem so I think you should do:

```
Destroy();
::new (static_cast<void*>(&other.m_value)) T(std::forward<U>(other)); // throws
m_hasValue = true;
```

(Note that this also requires fixing const U& to U&&)

**Comment by BillyONeal** [comment]
> > Wouldn't be better to have the implementation similar to `std::optional` and if C++ 17 is available, default to `std::optional`?
> > https://github.com/microsoft/STL/blob/master/stl/inc/optional
> 
> I think @BillyONeal is the best person to answer this question.

TL;DR: The template metaprogramming inside `std::optional` is extremely complex and I don't think we need to take that cost for the limited use cases we need to support in the SDK. If we decide we need the full power of that we should copy/paste an existing std::optional implementation rather than trying to build our own.

This whole thing started as what I considered the 'minimum viable product' for a type like this.

File: sdk/core/azure-core/inc/nullable.hpp [hunk 9]
```diff
@@ -103,6 +103,16 @@ namespace Azure { namespace Core {
       return *this;
     }
 
+    template <
+        class U = T,
+        typename std::enable_if<
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:108]: This constraint doesn't make sense because U isn't in the interface of operator= here.

File: sdk/core/azure-core/inc/nullable.hpp (hunk 6 above)
Comment (gilbertw) [reply sdk/core/azure-core/inc/nullable.hpp:148]: It made sense to me to explain why we abort.

Comment (BillyONeal) [reply sdk/core/azure-core/inc/nullable.hpp:148]: LGTM then :)

File: sdk/core/azure-core/inc/nullable.hpp (hunk 8 above)
Comment (gilbertw) [reply sdk/core/azure-core/inc/nullable.hpp]: Fixed

File: sdk/core/azure-core/inc/nullable.hpp [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+#pragma once
+
+#include <cstdlib> // for abort
+#include <new> // for placement new
+#include <type_traits>
+#include <utility> // for swap and move
+
+namespace Azure { namespace Core {
+  namespace Details {
+    struct NontrivialDummyType
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:13]: I think "Dummy" will trigger Policheck. How about just `EmptyType`

File: sdk/core/azure-core/inc/nullable.hpp [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+    {
+      constexpr NontrivialDummyType() noexcept {}
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:15]:
```suggestion
```

No need for a ctor, aggregates get constexpr ctors by default

File: sdk/core/azure-core/inc/nullable.hpp [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+    ~Nullable()
+    {
+      if (m_hasValue)
+      {
+        m_value.~T();
+      }
+    }
+
+    void Reset() noexcept /* enforces termination */ { Destroy(); }
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:67]: Technically the termination enforcement is happening in Destroy... OTOH this is the public one. Up to you.

File: sdk/core/azure-core/inc/nullable.hpp [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    template <
+        class U = T,
+        typename std::enable_if<
+            !std::is_same<Nullable, typename std::remove_cv<typename std::remove_reference<U>::type>::type>::value // Avoid repeated assignment
+            && !(std::is_scalar<U>::value && std::is_same<T, typename std::decay<U>::type>::value) //Avoid repeated assignment of equivallent scaler types
+            && std::is_constructible<T, U>::value   //Ensure the type is constructible
+            && std::is_assignable<T, U>::value,      //Ensure the type is assignable
+            int>::type
+        = 0>
+    Nullable& operator=(U&& other)
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:129]:
```suggestion
    Nullable& operator=(U&& other) noexcept(std::is_nothrow_constructible<T, U>::value && std::is_nothrow_assignable<T&, U>::value)
```

File: sdk/core/azure-core/inc/nullable.hpp [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+      return *this;
+    }
+
+    template <
+        class U = T,
+        typename std::enable_if<
+            !std::is_same<Nullable, typename std::remove_cv<typename std::remove_reference<U>::type>::type>::value // Avoid repeated assignment
+            && !(std::is_scalar<U>::value && std::is_same<T, typename std::decay<U>::type>::value) //Avoid repeated assignment of equivallent scaler types
+            && std::is_constructible<T, U>::value   //Ensure the type is constructible
+            && std::is_assignable<T, U>::value,      //Ensure the type is assignable
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:126]:
```suggestion
            && std::is_assignable<T&, U>::value,      //Ensure the type is assignable
```

File: sdk/core/azure-core/inc/nullable.hpp [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+            int>::type
+        = 0>
+    Nullable& operator=(U&& other)
+    {
+      Destroy();
+      ::new (static_cast<void*>(&m_value)) T(std::forward<U>(other));
+      m_hasValue = true;
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:133]:
```suggestion
      if (m_hasValue) {
        m_value = std::forward<U>(other);
      } else {
        ::new (static_cast<void*>(&m_value)) T(std::forward<U>(other));
        m_hasValue = true;
      }
```

File: sdk/core/azure-core/inc/nullable.hpp [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+      return *this;
+    }
+
+    template <class... U> T& Emplace(U&&... Args)
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:137]:
```suggestion
    template <class... U> T& Emplace(U&&... Args) noexcept(std::is_nothrow_constructible<T, U...>::value)
```

File: sdk/core/azure-core/inc/nullable.hpp [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+      {
+        // throwing here prohibited by our guidelines
+        // https://azure.github.io/azure-sdk/cpp_design.html#pre-conditions
+        std::abort();
+      }
+
+      return std::move(m_value);
+    }
+
+    explicit operator bool() const { return HasValue(); }
```
Comment (BillyONeal) [inline sdk/core/azure-core/inc/nullable.hpp:182]:
```suggestion
    explicit operator bool() const noexcept { return HasValue(); }
```

File: sdk/core/azure-core/inc/nullable.hpp (hunk 7 above)
Comment (BillyONeal) [reply sdk/core/azure-core/inc/nullable.hpp:184]: (If you don't add value_or you might want to delete this comment)

File: sdk/core/azure-core/inc/nullable.hpp (hunk 10 above)
Comment (gilbertw) [reply sdk/core/azure-core/inc/nullable.hpp:13]: Good catch, Fixed
